image.go

```diff
@@ -2,25 +2,177 @@ package llblib
+			// state stack.
+			cp, err := deep.Copy(cs.config)
+			if err != nil {
+				return llb.State{}, errtrace.Errorf("failed to copy image config: %w", err)
+			}
+			cs.config = cp
+		}
+		if cs.mutator == nil {
+			cs.mutator = m
+			return st.WithValue(imageConfigKey{}, cs), nil
+		}
+		// chain previous mutator to the one passed in
+		prev := cs.mutator
+		cs.mutator = func(ctx context.Context, c *mdispec.DockerOCIImage) error {
+			if err := prev(ctx, c); err != nil {
+				return errtrace.Wrap(err)
+			}
+			return errtrace.Wrap(m(ctx, c))
+		}
+		return st.WithValue(imageConfigKey{}, cs), nil
+	})
+}
+
 // ResolvedImage returns an llb.State where the image will be resolved with the
 // image configuration applied to the state.  The resolved image digest will
 // also be applied to the state to ensure this state is always consistent
-// during the solve execution.  Note that `WithImageResolver` should be used
-// on the context provided to the buildkit solve requests, otherwise an
-// un-resolved llb.Image will be used.
+// during the solve execution.
 func ResolvedImage(ref string, opts ...llb.ImageOption) llb.State {
+	return Image(ref, append(opts, llb.ResolveDigest(true))...)
+}
+
+// Image is similar to llb.Image but the image config will be preserved
+// so that the llb.State can be pushed to a registry.
+func Image(ref string, opts ...llb.ImageOption) llb.State {
 	return llb.Scratch().Async(func(ctx context.Context, _ llb.State, c *llb.Constraints) (llb.State, error) {
-		if resolver := LoadImageResolver(ctx); resolver != nil {
-			opts = append(opts,
-				llb.WithMetaResolver(resolver),
-				llb.ResolveDigest(true),
-			)
-			return llb.Image(ref, opts...), nil
+		capturingMetaResolver := &capturingMetaResolver{
+			resolver: LoadImageResolver(ctx),
 		}
-		return llb.Image(ref, opts...), nil
+
+		img := llb.Image(ref,
+			append(opts, llb.WithMetaResolver(capturingMetaResolver))...,
+		)
+		// this will un-async the image allowing us to capture the image config
+		img.Output().Vertex(ctx, c)
+		return img.Async(func(ctx context.Context, st llb.State, c *llb.Constraints) (llb.State, error) {
+			if len(capturingMetaResolver.config) == 0 {
+				return st, nil
+			}
+			var config mdispec.DockerOCIImage
+			if err := json.Unmarshal(capturingMetaResolver.config, &config); err != nil {
+				return llb.State{}, errtrace.Errorf("failed to unmarshal image config: %w", err)
+			}
+			return withImageConfig(st, &config), nil
+		}), nil
 	})
 }
+
+// imageResolverOption is a helper to extract the ImageMetaResolver if provided
+// on the llb.ImageOptions.
+func imageResolverOption(ctx context.Context, opts ...llb.ImageOption) sourceresolver.ImageMetaResolver {
+	ii := llb.ImageInfo{}
+	for _, opt := range opts {
+		opt.SetImageOption(&ii)
+	}
+	// HACK: the metaResolver field is private, so we have to use reflection to
+	// access it.  TODO maybe we can make this exported or add an accessor
+	// to llb.ImageInfo?
```

Comment:
Ugh, that is pretty bad.

Reply:
Yeah, definitely don't love this,  I added a few tests to verify it works.  We can easily unhack if buildkit makes this field available.